### PR TITLE
[Improve](array-func)improve array union support multi params

### PR DESCRIPTION
**Pull request by amorynan**
## Proposed changes

Issue Number: close #xxx
Now we just support array union with two arrays 

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

File: gensrc/script/doris_builtins_functions.py [hunk 1]
```diff
@@ -469,24 +469,24 @@
...
-        [['array_union'], 'ARRAY_DECIMAL64', ['ARRAY_DECIMAL64', 'ARRAY_DECIMAL64'], ''],
-        [['array_union'], 'ARRAY_DECIMAL128', ['ARRAY_DECIMAL128', 'ARRAY_DECIMAL128'], ''],
-        [['array_union'], 'ARRAY_DATETIME',  ['ARRAY_DATETIME', 'ARRAY_DATETIME'], ''],
-        [['array_union'], 'ARRAY_DATE',      ['ARRAY_DATE', 'ARRAY_DATE'], ''],
-        [['array_union'], 'ARRAY_DATETIMEV2',  ['ARRAY_DATETIMEV2', 'ARRAY_DATETIMEV2'], ''],
-        [['array_union'], 'ARRAY_DATEV2',      ['ARRAY_DATEV2', 'ARRAY_DATEV2'], ''],
-        [['array_union'], 'ARRAY_VARCHAR',   ['ARRAY_VARCHAR', 'ARRAY_VARCHAR'], ''],
-        [['array_union'], 'ARRAY_STRING',    ['ARRAY_STRING', 'ARRAY_STRING'], ''],
+        [['array_union'], 'ARRAY_BOOLEAN',  ['ARRAY_BOOLEAN', 'ARRAY_BOOLEAN', '...'],  ''],
+        [['array_union'], 'ARRAY_TINYINT',  ['ARRAY_TINYINT', 'ARRAY_TINYINT', '...'],  ''],
```
Comment (morrySnow) [inline gensrc/script/doris_builtins_functions.py:473]: should update `org.apache.doris.nereids.trees.expressions.functions.scalar.ArrayUnion` to let it support multi param in Nereids too

File: be/src/vec/functions/array/function_array_map.h [hunk 2]
```diff
@@ -76,7 +84,17 @@ struct OpenMapImpl {
             }
             // make map result to dst
             for (const auto& entry : map) {
-                if (entry.get_mapped() == params.size()) {
+                if constexpr (operation == MapOperation::INTERSECT) {
+                    if (entry.get_mapped() == params.size()) {
+                        ++dst_off;
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_map.h:89]: same code as line 99~103

File: be/src/vec/functions/array/function_array_map.h [hunk 3]
```diff
@@ -121,7 +139,17 @@ struct OpenMapImpl<operation, ColumnString> {
             }
             // make map result to dst
             for (const auto& entry : map) {
-                if (entry.get_mapped() == params.size()) {
+                if constexpr (operation == MapOperation::INTERSECT) {
+                    if (entry.get_mapped() == params.size()) {
+                        auto& dst_col = static_cast<ColumnString&>(*dst.nested_col);
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_map.h:144]: same code as line 153~159

File: be/src/vec/functions/array/function_array_union.cpp [hunk 4]
```diff
@@ -31,41 +32,46 @@ struct NameArrayUnion {
...
+    bool apply_null() { return result_null_flag; }
 
     // Handle Non-Null element.
-    // Return ture means this Non-Null element should put into result column.
-    template <bool is_left>
-    bool apply(Set& set, Set& result_set, const Element& elem) {
-        if (!set.find(elem)) {
-            set.insert(elem);
-            return true;
+    void apply(Map& map, const int arg_idx, const int row_idx,
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_union.cpp:48]: 1. row_idx better use 64bit instead int.
2. const int == int?

**Review by xy720 — COMMENTED** [review]
Why calculate multi array union need a map structure? Set is not enough?

**Comment by amorynan** [comment]
> Why calculate multi array union need a map structure? Set is not enough?

because MapImpl can apply const ColumnArrayExecutionDatas params, but SetImpl just apply one ColumnArrayExecutionData

File: be/src/vec/functions/array/function_array_map.h (hunk 2 above)
Comment (amorynan) [reply be/src/vec/functions/array/function_array_map.h:89]: I will make micro define

File: be/src/vec/functions/array/function_array_union.cpp (hunk 4 above)
Comment (amorynan) [reply be/src/vec/functions/array/function_array_union.cpp:48]: maybe size_t is better ?

**Comment by amorynan** [comment]
run buildall

File: be/src/vec/functions/array/function_array_union.cpp (hunk 4 above)
Comment (amorynan) [reply be/src/vec/functions/array/function_array_union.cpp]: done

**Review by morrySnow — DISMISSED** [review]

**Review by cambyzju — DISMISSED** [review]
LGTM

**Comment by amorynan** [comment]
run p0

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

**Comment by xiaokang** [comment]
run buildall

**Comment by amorynan** [comment]
run p0

**Comment by amorynan** [comment]
run p1

**Comment by amorynan** [comment]
@shuke987

**Review by morrySnow — APPROVED** [review]

**Review by xiaokang — APPROVED** [review]
LGTM

**Pull request merged**